alacritty_terminal/src/clipboard.rs

```diff
@@ -24,74 +25,139 @@ use copypasta::wayland_clipboard;
 use copypasta::x11_clipboard::{Primary as X11SelectionClipboard, X11ClipboardContext};
 use copypasta::{ClipboardContext, ClipboardProvider};
 
+use crate::event::{Event, EventListener};
+use crate::util;
+
 pub struct Clipboard {
+    _thread: std::thread::JoinHandle<()>,
+    tx: mpsc::Sender<ClipboardRequest>,
+}
+
+pub struct ClipboardStore {
     clipboard: Box<dyn ClipboardProvider>,
     selection: Option<Box<dyn ClipboardProvider>>,
 }
 
+pub enum ClipboardType {
+    Clipboard,
+    Selection,
+}
+
+enum ClipboardAction {
+    Store(String),
+    Load,
+}
+
+struct ClipboardRequest {
+    action: ClipboardAction,
+    ty: ClipboardType,
+}
+
 impl Clipboard {
     #[cfg(any(target_os = "macos", target_os = "windows"))]
     pub fn new() -> Self {
-        Self::default()
+        let clipboard = ClipboardStore {
+            clipboard: Box::new(ClipboardContext::new().unwrap()),
+            selection: None,
+        };
+
+        let (tx, rx) = mpsc::channel();
+        Self {
+            _thread: util::thread::spawn_named("clipboard", move || {
+                clipboard_thread(clipboard, rx, event_proxy);
+            }),
+            tx,
+        }
     }
 
     #[cfg(not(any(target_os = "macos", target_os = "windows")))]
-    pub fn new(display: Option<*mut c_void>) -> Self {
-        if let Some(display) = display {
+    pub fn new<T: EventListener + Send + 'static>(
+        display: Option<*mut c_void>,
+        event_proxy: T,
+    ) -> Clipboard {
+        let clipboard = if let Some(display) = display {
             let (selection, clipboard) =
                 unsafe { wayland_clipboard::create_clipboards_from_external(display) };
-            return Self { clipboard: Box::new(clipboard), selection: Some(Box::new(selection)) };
-        }
+            ClipboardStore { clipboard: Box::new(clipboard), selection: Some(Box::new(selection)) }
+        } else {
+            ClipboardStore {
+                clipboard: Box::new(ClipboardContext::new().unwrap()),
+                selection: Some(Box::new(
+                    X11ClipboardContext::<X11SelectionClipboard>::new().unwrap(),
+                )),
+            }
+        };
 
+        let (tx, rx) = mpsc::channel();
         Self {
```

Comment:
I've followed the implementation of our config watcher thread. In case we need some kind of graceful shutdown, let me know.